Dataset that was taken from HackerNews website which includes user submitted stories that is mainly generated the data-set. Info of post will be anlayzed to create insight.

In [3]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file) #read the data-set as list of lists

In [4]:
hn[:5] #first five rows

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

In [5]:
headers = hn[0] 
hn = hn[1:]
print(headers, end="\n")
print(hn[0:5]) #display header removed from dataset

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [6]:
ask_posts = [] #posts title with ask hn 
show_posts = [] #posts title with show hn 
other_posts = [] # other posts 
for row in hn:
    title = row[1] #title column in the dataset
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif  title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [7]:
total_ask_comments = 0
for row in ask_posts:
    comment = int(row[4]) #number of comment for post as integer
    total_ask_comments += comment

avg_ask_comment = total_ask_comments / len(ask_posts)
print("average ask comment:", avg_ask_comment)

total_show_comments = 0
for row in show_posts:
    s_comment = int(row[4])
    total_show_comments += s_comment
    
avg_show_comments = total_show_comments / len(show_posts)
print("average show comment:", avg_show_comments)

average ask comment: 14.038417431192661
average show comment: 10.31669535283993


Average number of comments for posts that ask something to Hacker News community is around 14.038 while Average number of comments for posts related with some findings or showcase to the community is 10.316.

In [8]:
import datetime as dt
result_list = [] #list of lists with hour of the post created and number of comment for a post

for row in ask_posts:
    post_hour = row[6]
    comment = int(row[4])
    result_list.append([post_hour, comment])
    
counts_by_hour = {} #number of ask posts created during each hour of the day
comments_by_hour = {} #number of comments ask posts created at each hour
for row in result_list:
    comment_date = row[0]
    comment_date = dt.datetime.strptime(comment_date, "%m/%d/%Y %H:%M")
    comment_date = comment_date.strftime("%H")
    if comment_date not in counts_by_hour:
        counts_by_hour[comment_date] = 1
        comments_by_hour[comment_date] = row[1]
    else:
        counts_by_hour[comment_date] += 1
        comments_by_hour[comment_date] += row[1]
        

In [9]:
avg_by_hour = [] # hour and average number of comments per post
for hour in counts_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour]])
    
avg_by_hour

[['13', 14.741176470588234],
 ['14', 13.233644859813085],
 ['17', 11.46],
 ['23', 7.985294117647059],
 ['08', 10.25],
 ['05', 10.08695652173913],
 ['15', 38.5948275862069],
 ['10', 13.440677966101696],
 ['12', 9.41095890410959],
 ['03', 7.796296296296297],
 ['22', 6.746478873239437],
 ['20', 21.525],
 ['04', 7.170212765957447],
 ['11', 11.051724137931034],
 ['07', 7.852941176470588],
 ['02', 23.810344827586206],
 ['19', 10.8],
 ['21', 16.009174311926607],
 ['16', 16.796296296296298],
 ['01', 11.383333333333333],
 ['06', 9.022727272727273],
 ['09', 5.5777777777777775],
 ['18', 13.20183486238532],
 ['00', 8.127272727272727]]

In [16]:
swap_avg_by_hour = [] #avg_by_hour with swapped columns
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [26]:
print("Top 5 Hours for Ask Posts Comments")
for row in sorted_swap[:5]:
    row[1] = str(row[1]) + ":00"
    print("{}: {:.2f} average comments per post.".format(row[1], row[0]))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post.
02:00: 23.81 average comments per post.
20:00: 21.52 average comments per post.
16:00: 16.80 average comments per post.
21:00: 16.01 average comments per post.


Above cell indicate most five hours that a post having higher chance of receiving comments.